fix #1536 (multiraft): drain (not close) channels on shutdown

tbg opened the pull request:
fixes #1536

tamird commented:
LGTM but ping @bdarnell because there should be a better way to deal with this :(

tschottdorf commented:
It's possible with a little bit of refactoring. With channel-based communication, the writer needs to close the channel, and that's the Store. We could have MultiRaft not use the Stopper: `NewMultiRaft` should return a closer chan for the `state` loop, and `(*Store).Close()` would close the closer, at which point MultiRaft doesn't even need to close its input channels any more.

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -623,22 +623,22 @@ func (s *state) stop() {
-			return
+	// on these operations. We can't close these channels unless we
+	// make sure that we never write to them after, but that would mean
+	// an atomic guard before every write access. Better to just deal
```
The only time multiraft writes to its own input channels is when it re-submits all pending proposals after an election; an atomic guard there is fine (and much better than deliberately leaking a goroutine to consume from the channels forever).

tschottdorf replied:
it's also about the store writing to the channels. Do we want an atomically updated bool in front of `CreateGroup`, `RemoveGroup` and `Propose` chans? And why is leaking the goroutine bad per se?

bdarnell replied:
On the store side all of these things should already be atomically guarded by `Stopper.StartTask`. 

Leaking a goroutine isn't a big problem since this won't happen more than once per process in production (although this implementation holds a reference to the `state` object and might indirectly keep a bunch of objects from GC), but it just seems clumsy to require it.

tschottdorf commented:
@bdarnell PTAL. I've guarded `store.ProposeRaftCommand` and `store.RemoveRange` so that they shouldn't try to use Raft when it may have shut down. I've removed the channel cleanup in `MultiRaft.close()` because nobody must be using the input channels any more, so nobody blocks. There's also no reason for closing them as far as I can tell.

bdarnell commented on storage/store.go:
```diff
@@ -1123,6 +1123,10 @@ func (s *Store) addRangeToRangeMap(rng *Range) error {
 func (s *Store) RemoveRange(rng *Range) error {
 	// RemoveGroup needs to access the storage, which in turn needs the
 	// lock. Some care is needed to avoid deadlocks.
+	if !s.stopper.StartTask() {
```
This looks like the wrong place for a call to `StartTask`; we need to ensure that a task is started at a higher level before calling whatever calls RemoveRange. I assume this is via `EndTransaction -> mergeTrigger -> mergeRange`, in which case I think the right place to call StartTask is in `Store.processRaft` right after we've read an event from the multiraft channel. Once we're draining we should just drop multiraft events on the floor; they'll be retried by the next incarnation of the node.

tschottdorf replied:
we still need to process Raft while draining, it's only right after having drained when Raft no longer needs to be able to make progress. I'll add that check to `processRaft`.

tschottdorf commented on storage/store.go:
```diff
@@ -1428,13 +1432,23 @@ func (s *Store) ProposeRaftCommand(idKey cmdIDKey, cmd proto.InternalRaftCommand
+	// If the system is shutting down and has drained, don't accept new
+	// proposals: Raft may already have stopped.
+	select {
+	case <-s.stopper.ShouldStop():
```
we also shouldn't be seeing this one do anything unless spurious requests exist, correct?

bdarnell replied:
Correct.

tbg committed:
- minor edits for StartTask usage (noop)
- remove multiraft close() channel massaging.
  it should not be necessary; use of the stopper guarantees
  that nothing waits on Raft any more.

tschottdorf commented:
I went back to check the commit this error came from, and `getLeaseForGossip` wasn't guarded by a `StartTask` call then. That means this race shouldn't be there now. I've kept the rest (including removing the MultiRaft channel dance which shouldn't be necessary).

bdarnell commented:
LGTM
